[WIP] Create DAGDependencyV2

enavarro51 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Creates a new DAGDependencyV2

### Details and comments

Previously `DAGDependency` was only used for the template matching transpiler pass. Additional uses are planned and there were several problems with the previous version.
- `DAGDependency` used `DAGDepNode` which stored a lot of information at the node level in attributes such as `successors`, `qindices`, and `matchedwith` that were only meant for use for template matching. These created excessive memory usage and did not take advantage of the rustworkx graph functions at runtime.
- The format and layout of the methods in `DAGDependency` had diverged significantly from `DAGCircuit`, even though they share a lot of functionality in common.
- The template matching code is slow and needs refactoring and a change to `DAGDependency` is part of that refactoring process.

The significant changes made are
- There is a new `DAGDependencyV2` which no longer uses `DAGDepNode` at all, but instead uses the existing `DAGOpNode` without modification. The makes for a cleaner and much reduced memory profile for the nodes.
- Additional methods from `DAGCircuit`, such as `find_bit` were added to the `DAGDependencyV2` and other methods were modified to more closely match the 2 dags.
- The information stored previously in the node attributes is now accessed directly in the template matching code through calls to the rustworkx functions.
- The template matching code has been modified to accomodate all of the above changes. This should result in a reduced memory profile, but at this stage may only result in a minor speed up.
- A terminology change was made in `DAGDependencyV2` and the template code. `direct_successors` is now `successors` and `successors` is now `descendants`. Similarly for `predecessors` and `ancestors`. This mirrors the usage in rustworkx.

For the future, a refactor of the template matching code is in order. One area especially is that the code mixes usage of the integer node id returned from rustworkx calls and the node object id, which results in many instances of having to map one to the other.

enavarro51 committed:
Initial changes

enavarro51 committed:
First matching changes

enavarro51 committed:
Debug

enavarro51 committed:
1

enavarro51 committed:
2

enavarro51 committed:
3

enavarro51 committed:
4

enavarro51 committed:
Create V2

enavarro51 committed:
Down to template_matching

enavarro51 committed:
Complete run on simple circuit and V2 cleanup

enavarro51 committed:
DAGDepNode bug

enavarro51 committed:
Add V2 converters

enavarro51 committed:
Full structure setup and more testing

enavarro51 committed:
Merge branch 'main' into new_dag_dep

enavarro51 committed:
Fix collect_blocks and collect_and_collapse

enavarro51 committed:
Get qindices working - 2 failures

enavarro51 committed:
Fixing qindices

enavarro51 committed:
Finish qindices_map and cindices_map

enavarro51 committed:
Fix final template code

enavarro51 committed:
Add cliffors tests back in

enavarro51 committed:
Fix soft compare bug

enavarro51 committed:
Cleanup docs and comments and add test_dagdependencyV2

enavarro51 committed:
Cleanup dag drawer

enavarro51 committed:
Merge branch 'main' into new_dag_dep

enavarro51 committed:
Lint and docs

enavarro51 committed:
Merge branch 'new_dag_dep' of github.com:enavarro51/qiskit-terra into new_dag_dep

enavarro51 committed:
Lint

enavarro51 committed:
Merge branch 'main' into new_dag_dep

enavarro51 committed:
Lint and change modules to _v2

enavarro51 committed:
Circular import

enavarro51 committed:
Remove merge_no_dupes

enavarro51 committed:
Add count ops

alexanderivrii commented on qiskit/converters/circuit_to_dagdependency_v2.py:
```diff
@@ -0,0 +1,45 @@
+
+    for instruction in circuit.data:
+        dagdependency.add_op_node(instruction.operation, instruction.qubits, instruction.clbits)
+
+    dagdependency.calibrations = circuit.calibrations
```
This seems an oversight in the existing code as well, but shouldn't we also update the `global_phase` (as is done in `dag_to_dagdependency_v2` code)?

enavarro51 replied:
This goes to your work to coordinate the 3 classes, which involves `add_calibration` as well. Do these things belong in all 3 classes? And if so, can they be implemented the same way?

alexanderivrii commented on qiskit/dagcircuit/collect_blocks.py:
```diff
@@ -100,12 +100,12 @@ def _direct_preds(self, node):
                     self.dag.get_node(pred_id)
-                    for pred_id in self.dag.direct_successors(node.node_id)
+                    for pred_id in self.dag.get_successors(self.dag.node_map[node])
                 ]
```
I have not yet looked at the internals of `DAGDependencyV2`, but getting the direct predecessors still significantly differs between `DAGCircuit` and `DAGDependencyV2`.

Using DAGCircuit:
```
[pred for pred in self.dag.predecessors(node) if isinstance(pred, DAGOpNode)]
```
Using DAGDependencyV2: 
```
[self.dag.get_node(pred_id) for pred_id in self.dag.get_successors(self.dag.node_map[node])]
```

I am wondering if we could bring the two expressions closer.

enavarro51 replied:
See `_node_id` reply below.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
```
Ha! This code was written at least 3 years ago. :)

enavarro51 replied:
Actually almost 4 years ago, mostly by Romain Moyard, an author of the template matching paper.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+        """
+        self._calibrations = defaultdict(dict, calibrations)
+
+    def add_calibration(self, gate, qubits, schedule, params=None):
+        """Register a low-level, custom pulse definition for the given gate.
```
This code is adapted from `DAGCircuit`, correct? A naive question: why should `DAGDependencyV2` support this method?

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+        """Returns the DAGDependencyV2 in rustworkx format."""
+        return self._multi_graph
+
+    def size(self):
+        """Returns the number of gates in the circuit"""
```
Interesting. The `size` function in `DAGCircuit` has an additional argument `recurse`, while the `size` function in `QuantumCircuit` has an additional argument `filter_function`. Do we want to extend this function here and how?

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+        self.comm_checker = CommutationChecker()
+
+        # Map of node to node index
+        self.node_map = {}
```
DAGNode/DAGOpNode have the field `_node_id`, which is what is used by `DAGCircuit` to map from a node to its rustworkx id. Can we use a similar strategy here instead of supporting `node_map`?

enavarro51 replied:
Somehow I completely missed that `_node_id` was used throughout DAGCircuit. I thought it was only used minimally, so I didn't use it in DAGDependencyV2, but should have. Since it would also need to be used in the template matching code and the underscore implies some level of privacy, we'd probably want to add an `@property` for `node_id` to the DAGDependencyV2 code.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+        ]
+
+    def get_in_edges(self, node_id):
+        """
```
A general comment/question about API. In this function and others, `node_id` is the "internal" integer id assigned by rustworkx graph, correct? Would it make sense to change the API to refer to DAGNodes (or DAGOpNodes) but not to these rustworkx ids? Is this how we already do this in `DAGCircuit`?

enavarro51 replied:
The reason I kept it this way currently is that is the way the template matching code uses it. I should probably do a second run through all this code to convert to using `node.node_id` as discussed above. This should simplify a lot of the code.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,725 @@
+        dag.qregs = self.qregs.copy()
+
+        for node in self.get_nodes():
+            dag._multi_graph.add_node(node.copy())
```
When copying dag dependencies, do we always need to copy the actual nodes too? As far as I can see, `DAGCircuit` does not have the function `copy` (and maybe it should). Ok, I guess we had to copy the nodes in the old template matching code where a lot of additional information was stored on nodes (and likely this information had to change in different parts of the algorithm), but maybe now (with this additional information stripped away) we can avoid this additional copying? 

Update: looking at template matching code, I see that you are ahead of this and are no longer copying DAGDependencies. Do we still need this `copy` function?

enavarro51 replied:
I actually removed all the dag_dep copying in the template code, so the copying here is only if a user wants to use it. We definitely do not have to copy the node here.

alexanderivrii commented on qiskit/transpiler/passes/optimization/template_matching/forward_match.py:
```diff
@@ -79,49 +79,49 @@ def __init__(
         # Transformation of the carg indices of the circuit to be adapted to the template indices
         self.carg_indices = []
 
+        self.successorstovisit = {}
+        self.isblocked = {}
+        self.matchedwith = {}
+
```
Great! So these are now used instead of the information that used to be stored on DAGDepNodes. This is much much cleaner.

enavarro51 replied:
The only leftover that is not purely in the template code is `qindices` and `cindices` which are computed and stored at the DAGDependencyV2 level. It's done in the `add_op_node` method. It's just a list of indices for the qargs and cargs for a node. It seemed easiest to store this at this level, rather than having to rebuild the list each time it's needed for a node in the template code. This can probably be optimized with a template code rework.

alexanderivrii reviewed (commented):
So far I only took a superficial look at the PR and I am definitely excited about the changes: DAGDependencyV2 is on the road to become cleaner, more efficient and more functional than DAGDependency, and the classes DAGCircuit and DAGDependencyV2 start looking closer to each other. I have left a few minor in-place comments and questions, but this is not a complete review.

What significantly confuses me (and I mean qiskit code in general, not particularly this PR) is the divergence of similar methods between QuantumCircuit, DAGCircuit and DAGDependency. As the simplest example, the `size` function in QuantumCircuit takes the argument `filter_function`, the `size` function in DAGCircuit takes the argument `recurse`, and the size function in `DAGDependency` does not take either (the same is true for `DAGDependencyV2`). Not only the `size` function should be used differently for different representations of the quantum circuit, it's likely to return different results depending on the representation. And this is just the tip of the iceberg. It would probably be nice to have an in-depth discussion about these 3 APIs and try to find the best solution. 

A related point is that the API for DAGDependency and (as of right now) DAGDependencyV2 is based on integer node ids assigned by rustworkx graphs, while the API for DAGCircuit is based on DAGNodes.  It would probably be good to have the DAGDependencyV2 API also in terms of nodes and not rustworkx ids.

enavarro51 committed:
Merge branch 'main' into new_dag_dep

enavarro51 committed:
Switch to using _node_id

enavarro51 committed:
Make more similar to DAGCircuit and add successor_indices, etc

enavarro51 commented:
@alexanderivrii I did another pass through `DAGDependencyV2` and the method calls now are a much closer match to the calls in `DAGCircuit`. See https://github.com/Qiskit/qiskit/pull/11310/commits/ddafcdeccc54657fdaaea496eb89bb66d599399e.

I added 4 new methods - `successor_indices, predecessor_indices, descendant_indices, and ancestor_indices` which are needed for the template matching code since it uses the indices primarily. In the future, if that code is refactored to use node objects instead of indices, these 4 can be removed from `DAGDependencyV2`.

I also removed several edge-related functions, such as `add_edge` and `get_edges` since these were not in `DAGCircuit` and I assume to maintain the integrity of the commute relations, users should not be doing direct edge manipulation.

enavarro51 committed:
Fix backward compat for collect_blocks

enavarro51 committed:
Update collect_block tests for both dag deps

enavarro51 committed:
Remove dag_dep remove methods & update dag viz

enavarro51 committed:
Add global_phase to dagdependency_v2 converters
